tutorials/deployment-with-docker.md

```diff
@@ -0,0 +1,119 @@
+# Introduction
+> Docker is a platform for developers and sysadmins to develop, deploy, and run applications with containers. The use of Linux containers to deploy applications is called containerization. Containers are not new, but their use for easily deploying applications is.
+> 
+> Containerization is increasingly popular because containers are:
+> 
+>   - Flexible: Even the most complex applications can be containerized.
+>   - Lightweight: Containers leverage and share the host kernel.
+>   - Interchangeable: You can deploy updates and upgrades on-the-fly.
+>   - Portable: You can build locally, deploy to the cloud, and run anywhere.
+>   - Scalable: You can increase and automatically distribute container replicas.
+>   - Stackable: You can stack services vertically and on-the-fly.
+>
+> Source [Get Started - docs.docker.com](https://docs.docker.com/get-started/) (I highly recommend to read into it if you didn't already)
+
+Docker provides alot of useful features for NodeCG:
+ - No need to care about dependencies they are all bundled with the docker image
+ - Run it everywhere.
+ - Easy upgrades.
+
+# Prerequirements
+- [Docker](https://docs.docker.com/install/) (I was using docker version `19.03.1-ce, build 74b1e89e8a`)
+- [Docker-Compose](https://docs.docker.com/compose/install/) (`1.13.0+`) (optional)
+
+# Bundle development
+If you want to use NodeCG in development, you can easily deploy a NodeCG instance with Docker. All you have to do is run:
+
+```sh
+docker run -it -u root -p 9090:9090 -rm nodecg/nodecg
+```
+> The arguments `-u root` is needed to ensure that the NodeCG process runnning inside the container has the rights to read the mounted volumes.
+
+To include the bundle your working on in this Docker Container. You can mount it like this:
+
+```sh
+docker run -it -u root -p 9090:9090 -rm -v [bundle-dir]:/opt/nodecg/bundles/[bundle-name] -v [config-file]:/opt/nodecg/cfg/[bundle].json nodecg/nodecg
+```
+> The argument `-u root` is needed to ensure that the NodeCG process runnning inside the container has the rights to read the mounted volumes.
+
+Mounting the volumes and running as `root` will ensure that the changes are in the container, without a docker image rebuild.
+
+# Production
+If you want to use NodeCG in production, it is recommended to build a own Docker Image to hold all your bundles and configuration.
+
+Example Dockerfile:
+```Dockerfile
+FROM nodecg/nodecg:v1.5.0
+
+RUN nodecg install [bundle-repo] && nodecg defaultconfig [bundle-name] && \
+    ... 
+    
+COPY --chown=nodecg:nodecg [nodecg-config-file] /opt/nodecg/cfg/nodecg.json
+COPY --chown=nodecg:nodecg [bundle-config-file] /opt/nodecg/cfg/[bundle].json
```

Comment:
You should never include config files in Docker image. It sometimes contain sensitive information like credentials.
Even in production it should be served through volumes.

Reply:
In Docker for Windows mounting that volume could result in a denied file access.

Reply:
I don't main Windows, but I assume it's still fixable by managing file permission...? Would it justify including sensitive or environment-specific information inside Docker images?

Reply:
The last time I checked it was not possible to mount a file with specific permissions in Docker for Windows. It would require running the image as root.
I dont think adding configuration to a private docker image is a big problem. It would be an idea to let the user use environment variables to overwrite configuration. That would take either a entrypoint-script that reads the environment variables at startup and write them to the files or some changes to the nodecg config module (`convict` should allow that in a very easy way).

Reply:
Fair enough. It should at least be optional and mention that this might be required when running on Windows and if doing it the resulting image should never be pushed to the registry.
It might be too much, but as a tutorial starting from "what is Docker", I'd like to be extra careful.